README.md
## AWS Ci Cd Fargate

A library that creates a full out-of-the-box solution for ECS Fargate with CI/CD pipeline.

#### Remarks

The project is written by [Deividas Tamkus](https://github.com/deitam), supervised by 
[Laimonas Sutkus](https://github.com/laimonassutkus) and is owned by 
[iDenfy](https://github.com/idenfy). This is an open source
library intended to be used by anyone. [iDenfy](https://github.com/idenfy) aims
to share its knowledge and educate market for better and more secure IT infrastructure.

#### Related technology

This project utilizes the following technology:

- *AWS* (Amazon Web Services).
- *AWS CDK* (Amazon Web Services Cloud Development Kit).
- *AWS CloudFormation*.
- *AWS Loadbalancer*.
- *AWS ECS* (Amazon Web Services Elastic Container Service).
- *AWS Fargate* (Serverless solution for ECS).
- *AWS CodePipeline*.

#### Assumptions

This library project assumes the following:

- You have knowledge in AWS (Amazon Web Services).
- You have knowledge in AWS CloudFormation and AWS loadbalancing.
- You are managing your infrastructure with AWS CDK.
- You are writing AWS CDK templates with a python language.

#### Install

The project is built and uploaded to PyPi. Install it by using pip.

```bash
pip install aws-fargate-cdk
```

Or directly install it through source.

```bash
./build.sh -ic
```

#### Description

This package creates a Fargate service with autoscaling, balancing and two pipelines 
for a complete out-of-the-box hosting infrastructure.

The pipelines are as follows:

1. **ECR to ECS**. This pipeline takes an image pushed to ECR and deploys it to Fargate using Blue/Green deployment.
The pipeline needs to be triggered manually duo to AWS CloudWatch event bugs related to ECR.
2. **CodeCommit to ECR**. This pipeline takes code pushed to the master branch of a CodeCommit repository, builds an image out of it (_source code needs a Dockerfile_), pushes it to ECR and automatically triggers the first pipeline, which then deploys it to ECS.

**TL;DR** Pushing source code with a Dockerfile to CodeCommit repository deploys it to ECS Fargate.

#### Examples

Create a fargate service with ci/cd:

```python
ecs_params = EcsParams(...)
load_params = LoadBalancerParams(...)
pipeline_params = PipelineParams(...)
listener_params = LbListenerParameters(...)

EcsFargateWithCiCd(
    scope=scope,
    prefix='pre',
    vpc=vpc,
    lb_params=load_params,
    ecs_params=ecs_params,
    lb_listener_params=listener_params,
    pipeline_params=pipeline_params
)
```
#### Tutorial

- Create a full infrastructure around ECS Fargate by using the following code below in your stack.

```python
from aws_cdk import core, aws_ec2, aws_elasticloadbalancingv2
from aws_ci_cd_fargate.parameters.ecs_parameters import EcsParams
from aws_ci_cd_fargate.parameters.pipeline_parameters import PipelineParams
from aws_ci_cd_fargate.parameters.load_balancer_parameters import LoadBalancerParams
from aws_ci_cd_fargate.parameters.lb_listener_parameters import LbListenerParameters
from aws_ci_cd_fargate.ecs_fargate_with_ci_cd import EcsFargateWithCiCd

class MainStack(core.Stack):
    def __init__(self, scope: core.App) -> None:
        super().__init__(
            scope=scope,
            id='MyCoolStack'
        )

        # Create your own vpc or use an existing one.
        vpc = aws_ec2.Vpc(...)
        
        # Create a security group for your ECS Fargate instances.
        sg = aws_ec2.SecurityGroup(...)
        
        # Create a loadbalancer.
        loadbalancer = aws_elasticloadbalancingv2.ApplicationLoadBalancer(...)
        production_listener = aws_elasticloadbalancingv2.ApplicationListener(self, 'Prod', load_balancer=loadbalancer)
        deployments_listener = aws_elasticloadbalancingv2.ApplicationListener(self, 'Test', load_balancer=loadbalancer)
        
        ecs_params = EcsParams('FargateEcsContainer', 256, 512, 80, {}, [sg], vpc.private_subnets)
        load_params = LoadBalancerParams()
        pipeline_params = PipelineParams()
        listener_params = LbListenerParameters(
            production_listener=production_listener,
            deployment_listener=deployments_listener,
            rule_priority=100,
            rule_condition=aws_elasticloadbalancingv2.CfnListenerRule.RuleConditionProperty(
                    field='path-pattern',
                    path_pattern_config=aws_elasticloadbalancingv2.CfnListenerRule.PathPatternConfigProperty(
                        values=['/*']
                    )
                )
        )

        self.ecs_infrastructure = EcsFargateWithCiCd(
            scope=self,
            prefix='MyCool',
            vpc=vpc,
            lb_params=load_params,
            ecs_params=ecs_params,
            lb_listener_params=listener_params,
            pipeline_params=pipeline_params
        )
        
        # Access CodeCommit-To-Ecr pipeline.
        _ = self.ecs_infrastructure.pipeline.commit_to_ecr
        
        # Access Ecr-To-Ecs pipeline.
        _ = self.ecs_infrastructure.pipeline.ecr_to_ecs
```

- Provision you infrastructure with `CloudFormation` by calling `cdk deploy`.

- Create a Dockerfile as simple as:

```dockerfile
FROM nginx
```

- After you provision your infrastructure, go to `AWS CodeCommit` in your AWS Console.

- Find a newly created git repository.

- Commit the Dockerfile to the newly created repository to trigger a pipeline.

(A tutorial on pushing code to remote repositories: [AWS Tutorial](https://docs.aws.amazon.com/codecommit/latest/userguide/how-to-create-commit.html)).

(A tutorial on setting up git ssh with aws git repositories: [AWS Tutorial](https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-ssh-unixes.html))


HISTORY.md
# Release history

#### 7.3.0
Add md files.

#### 7.2.0
AWS CDK dependency update 1.60.0 - 2.0.0.

#### 7.1.0
Force dependency update 1.44.0 AWS CDK.

#### 7.0.0
Do not allow to specify ecs port. Default to 80.
This is AWS limitation.

#### 6.0.0
Instead of taking path config as a parameter, take the 
whole rule config.

#### 5.0.1
Use empty ecs repository dependency.

#### 5.0.0
Project rename to a better and more explanatory name.

#### 4.0.9
Dependency update.

#### 4.0.8
Dependency update. Fix ecs service name key.

#### 4.0.7
Dependency update.

#### 4.0.6
Use newest cdk with newest breaking changes.

#### 4.0.5
Add dependency which creates ecs service. CloudFormation itself
has too many bugs.

#### 4.0.2
Destroy ecr on delete.

#### 4.0.0
Build environment variable fix.

#### 4.0.0
Add pipeline parameters and fix some major pipeline bugs.

#### 3.1.0
Create custom resource to create ecs service. With deployment controller CODE_DEPLOY you
can not do CF updates.

#### 3.0.2
Minor bug fix.

#### 3.0.1
Readme fix.

#### 3.0.0
Full project refactor. Accept loadbalancer's listeners for production and test traffic instead of creting
a loadbalancer here. This way we can reuse an existing loadbalancer.

#### 2.0.0
Remove pipeline parameters as the artifacts bucket should be created automatically within this stack.

#### 1.0.0
Initial project.


hello-cdk/hello_cdk/hello_cdk_stack.py
from aws_cdk import core, aws_ecs, aws_codedeploy, aws_codecommit, aws_codebuild, aws_codepipeline, \
    aws_codepipeline_actions, aws_ecr, aws_ecs_patterns, aws_elasticloadbalancingv2, aws_ec2, aws_certificatemanager, aws_iam


class MySubStack(core.Stack):

    TARGET_GROUP_PORT = 80
    LISTENER_HTTP_PORT_1 = 80
    LISTENER_HTTPS_PORT_1 = 443
    LISTENER_HTTP_PORT_2 = 8000
    LISTENER_HTTPS_PORT_2 = 44300

    def __init__(self, scope: core.Construct, id: str, desired_domain_name: str) -> None:
        super().__init__(scope, id)

        vpc = aws_ec2.Vpc(self, 'vpc')

        certificate = aws_certificatemanager.Certificate(self, 'cert', domain_name=desired_domain_name, validation_method=aws_certificatemanager.ValidationMethod.DNS)

        load_balancer = aws_elasticloadbalancingv2.ApplicationLoadBalancer(
            self, 'load_balancer', vpc=vpc
        )
        http_listener_1 = load_balancer.add_listener(
            'listener1',
            port=self.LISTENER_HTTP_PORT_1,
            protocol=aws_elasticloadbalancingv2.ApplicationProtocol.HTTP
        )
        http_listener_1.add_redirect_response(
            'https-redirect',
            host='#{host}',
            path='/#{path}',
            port=str(self.LISTENER_HTTPS_PORT_1),
            query='#{query}',
            status_code='HTTP_301',
            protocol='HTTPS'
        )

        http_target_group_1 = aws_elasticloadbalancingv2.ApplicationTargetGroup(
            self, 'target_group1',
            port=self.TARGET_GROUP_PORT,
            vpc=vpc,
            protocol=aws_elasticloadbalancingv2.ApplicationProtocol.HTTP,
            target_type=aws_elasticloadbalancingv2.TargetType.IP,
            health_check=aws_elasticloadbalancingv2.HealthCheck()
        )

        https_listener_1 = load_balancer.add_listener(
            'listener2',
            port=self.LISTENER_HTTPS_PORT_1,
            certificate_arns=[certificate.certificate_arn],
            default_target_groups=[http_target_group_1],
            protocol=aws_elasticloadbalancingv2.ApplicationProtocol.HTTPS
        )

        http_listener_2 = load_balancer.add_listener(
            'listener3',
            port=self.LISTENER_HTTP_PORT_2,
            protocol=aws_elasticloadbalancingv2.ApplicationProtocol.HTTP
        )

        http_listener_2.add_redirect_response(
            'https-redirect',
            host='#{host}',
            path='/#{path}',
            port=str(self.LISTENER_HTTPS_PORT_2),
            query='#{query}',
            status_code='HTTP_301',
            protocol='HTTPS'
        )

        http_target_group_2 = aws_elasticloadbalancingv2.ApplicationTargetGroup(
            self, 'target_group2',
            port=self.TARGET_GROUP_PORT,
            vpc=vpc,
            protocol=aws_elasticloadbalancingv2.ApplicationProtocol.HTTP,
            target_type=aws_elasticloadbalancingv2.TargetType.IP,
            health_check=aws_elasticloadbalancingv2.HealthCheck()
        )
        https_listener_2 = load_balancer.add_listener(
            'listener4',
            port=self.LISTENER_HTTPS_PORT_2,
            certificate_arns=[certificate.certificate_arn],
            default_target_groups=[http_target_group_2],
            protocol=aws_elasticloadbalancingv2.ApplicationProtocol.HTTPS
        )

        task = aws_ecs.FargateTaskDefinition(self, "Task", cpu=256, memory_limit_mib=512, family='myfamily')
        container = task.add_container("Container", image=aws_ecs.ContainerImage.from_registry('nginx:latest'))
        container.add_port_mappings(aws_ecs.PortMapping(container_port=80))
        service = aws_ecs_patterns.ApplicationLoadBalancedFargateService(
            self, "Fargate",
            cpu=256,
            memory_limit_mib=512,
            task_definition=task,
            load_balancer=load_balancer,
            assign_public_ip=True
        )
        scaling = service.service.auto_scale_task_count(max_capacity=5)
        scaling.scale_on_cpu_utilization("CpuScaling", target_utilization_percent=75)

        myapp = aws_codedeploy.CfnApplication(
            self, 'myapp',
            application_name='fargateapp',
            compute_platform='ECS'
        )

        deployment_role = aws_iam.Role(
            self, 'deploymentRole',
            path='/',
            inline_policies={'firstpolicy': aws_iam.PolicyDocument(
                statements=[aws_iam.PolicyStatement(
                    actions=[
                        "ecs:DescribeServices",
                        "ecs:CreateTaskSet",
                        "ecs:UpdateServicePrimaryTaskSet",
                        "ecs:DeleteTaskSet",
                        "elasticloadbalancing:DescribeTargetGroups",
                        "elasticloadbalancing:DescribeListeners",
                        "elasticloadbalancing:ModifyListener",
                        "elasticloadbalancing:DescribeRules",
                        "elasticloadbalancing:ModifyRule",
                        "lambda:InvokeFunction",
                        "cloudwatch:DescribeAlarms",
                        "sns:Publish",
                        "s3:GetObject",
                        "s3:GetObjectMetadata",
                        "s3:GetObjectVersion",
                        "iam:PassRole"
                    ],
                    resources=['*'],
                    effect=aws_iam.Effect.ALLOW)]
            )},
            assumed_by=aws_iam.ServicePrincipal('ecs-tasks.amazonaws.com')
        )

        aws_codedeploy.CfnDeploymentGroup(
            self, 'DeploymentGroup',
            service_role_arn=deployment_role.role_arn,
            application_name=myapp.application_name,
            deployment_group_name='mydeploymentgroup',
            deployment_config_name='CodeDeployDefault.ECSAllAtOnce'
        )

        myrepo = aws_codecommit.Repository(self, "Repository", repository_name="my_repository")
        source_artifact = aws_codepipeline.Artifact()
        build_artifact = aws_codepipeline.Artifact("BuildArtifact")
        project = aws_codebuild.Project(
            self, "Project",
            source=aws_codebuild.Source.code_commit(repository=myrepo),
            build_spec=aws_codebuild.BuildSpec.from_source_filename('push-to-ecr.yml')
        )

        pipeline = aws_codepipeline.Pipeline(self, "Pipeline")

        source = pipeline.add_stage(stage_name="Source")
        source.add_action(
            aws_codepipeline_actions.CodeCommitSourceAction(
                branch='master',
                repository=myrepo,
                output=source_artifact,
                action_name="Commit"
            )
        )

        build = pipeline.add_stage(stage_name="Build")
        build.add_action(
            aws_codepipeline_actions.CodeBuildAction(
                input=source_artifact,
                project=project,
                action_name="Build",
                outputs=[build_artifact]
            )
        )

        secondary_myrepo = aws_ecr.Repository(self, "EcrRepo")
        secondary_source_artifact = aws_codepipeline.Artifact()
        secondary_build_artifact = aws_codepipeline.Artifact("BuildArtifact")
        secondary_project = aws_codebuild.Project(
            self, "BuildProject",
            build_spec=aws_codebuild.BuildSpec.from_object({'deploy': 'deploy'})
        )

        secondary_pipeline = aws_codepipeline.Pipeline(self, "SecPipeline")

        secondary_source = secondary_pipeline.add_stage(stage_name="SecSource")
        secondary_source.add_action(
            aws_codepipeline_actions.EcrSourceAction(
                output=secondary_source_artifact,
                repository=secondary_myrepo,
                action_name="Source"
            )
        )

        secondary_build = secondary_pipeline.add_stage(stage_name="SecBuild")
        secondary_build.add_action(
            aws_codepipeline_actions.CodeBuildAction(
                input=secondary_source_artifact,
                project=secondary_project,
                action_name="Build", outputs=[secondary_build_artifact]
            )
        )

        secondary_deploy = secondary_pipeline.add_stage(stage_name="SecDeploy")
        secondary_deploy.add_action(
            aws_codepipeline_actions.EcsDeployAction(
                service=service.service,
                action_name="Deploy",
                input=build_artifact
            )
        )
